### PR TITLE
Support the reset of the local development environment to the snapshot state.

### PR DESCRIPTION
**Pull request by emetsger**
Adds make reset target which sets the system state to the current snapshot.

### CODE REVIEW

File: idc.Makefile [hunk 1]
```diff
@@ -48,6 +48,30 @@ snapshot-image:
 	$(MAKE) docker-compose.yml
 	docker-compose up -d
 
+.PHONY: reset
+.SILENT: reset
+reset: warning-destroy-state destroy-state
+	@echo "Removing vendored modules"
+	-rm -rf codebase/vendor
```
Comment (birkland) [inline idc.Makefile:55]: If the intent is to blow away modules & dependencies, and allow composer to re-populate then the following directories should also be emptied:
```
codebase/web/core/
codebase/web/modules/contrib/
codebase/web/themes/contrib/
```

**Review by birkland — DISMISSED** [review]
Could you update `IDC.md` to add this to the make targets, and explain what it does?

Also, I'm wondering if this should also do a `config-import` since it already blows away any database state, and we know for certain that there is no danger in doing so.   In fact, depending on the intention of this operation, there may be danger in running Drupal without loading our config (instead, relying on whatever older config might be in the snapshot image).

**Comment by emetsger** [comment]
> Also, I'm wondering if this should also do a `config-import` since it already blows away any database state, and we know for certain that there is no danger in doing so. In fact, depending on the intention of this operation, there may be danger in running Drupal without loading our config (instead, relying on whatever older config might be in the snapshot image).

The intent here is to use the configuration present in the snapshot, which is a reasonable thing to do, because embedding them in the snapshot is the mechanism by which they are provided to our colleagues for their use.  That is, creating a snapshot is an intentional act, with the intention of sharing any configuration settings it contains.  If this make target overwrites the settings shared by developers, how are they tested?  At any rate, any two developers should be able to perform a `make reset` and end up with the same state in their individual environments.

Overwriting the settings in the snapshot obviates that intent, and IMO if that is the intent of the developer, then they should be responsible for importing their settings.  Note that none of our `make` targets perform a `config-import` on behalf of the developer.  That is, we take care to _not_ overwrite active configuration; instead, it is an intentional action taken by the developer.

**Comment by birkland** [comment]
I see.  This does point out an eventual need for a "start up just like production would" target that _does_ do a sync.  At the end of the day, the sync'd config is the only config that matters, since it is the only config to be shipped off to the cloud instances.  The cloud instances will always overwrite their existing config at startup , and the developers will want to observe that happen at startup on their own machines too.   That being said, I do agree that there are situations where a developer might want to assess before and after behaviour of updating config values, and this target would facilitate that kind of investigation.   Thus, I'm fine with "do not import config" being an intentional part of this target.

**Review by birkland — CHANGES_REQUESTED** [review]
One issue related to snapshots:  The `reset` target does a pull.  If you have made a `snapshot-image`, but have not pushed the snapshot image, then this target will fail.  Is there a reason for the `pull`?  Can it be removed?    Maybe another way of asking this question:  Does running this target inherently presume you're resetting to an externally-accessible state (all images pushed), explicitly disallowing reset to an "internal" checkpoint snapshot you've made (but not pushed)?  

Minor thing:  In the documentation for `make reset`, mention that it will not pull in (sync) the configuration committed in `codebase`, and will use the configuration present on the snapshot image only.  

Minor thing:  You need to type in a capital Y at the yes/no question, a lower case ends in a make error.  Not sure if that's intentional or not

**Comment by emetsger** [comment]
> One issue related to snapshots: The `reset` target does a pull. If you have made a `snapshot-image`, but have not pushed the snapshot image, then this target will fail. Is there a reason for the `pull`? 

Yes, to insure that the images are pulled :)  If you publish a PR and tell people to do a `make reset`, they won't have the snapshot image unless a pull is performed.

> Can it be removed? Maybe another way of asking this question: Does running this target inherently presume you're resetting to an externally-accessible state (all images pushed), explicitly disallowing reset to an "internal" checkpoint snapshot you've made (but not pushed)?

I'm not making any statements about what is explicitly disallowed, I'm just saying that the intent is that a snapshot has been made, pushed, and when the code is shared, `make reset` can be performed to  pull in the latest snapshot.  There's not an intent to explicitly disallow someone from doing something, but it may be a consequence of this approach.

> 
> Minor thing: In the documentation for `make reset`, mention that it will not pull in (sync) the configuration committed in `codebase`, and will use the configuration present on the snapshot image only.

Sure.  Note that none of our targets pull in (sync) configuration.  So this target behaves as all of our other targets.

> 
> Minor thing: You need to type in a capital Y at the yes/no question, a lower case ends in a make error. Not sure if that's intentional or not

yes, intentional.  That's why it's a capital Y :)

**Comment by birkland** [comment]
> > One issue related to snapshots: The `reset` target does a pull. If you have made a `snapshot-image`, but have not pushed the snapshot image, then this target will fail. Is there a reason for the `pull`?
> 
> Yes, to insure that the images are pulled :) If you publish a PR and tell people to do a `make reset`, they won't have the snapshot image unless a pull is performed.

Ah sorry.  What I'm referring to `docker-compose`'s default "pull if an image is not present locally" behavior.  Relying on that, an explicit `docker-compose pull` is not necessary, right?   So I suppose I'm asking:  why explicitly `pull` rather than rely in the implicit "pull if not present" inherent in `docker-compose up -d`? 

As a test, try removing the all snapshot images from your docker environment, remove the `pull` steps from the `reset`  target,  run `make reset`, notice that the snapshot image is implicitly pulled, and notice all is good.

> > Can it be removed? Maybe another way of asking this question: Does running this target inherently presume you're resetting to an externally-accessible state (all images pushed), explicitly disallowing reset to an "internal" checkpoint snapshot you've made (but not pushed)?
> 
> I'm not making any statements about what is explicitly disallowed, I'm just saying that the intent is that a snapshot has been made, pushed, and when the code is shared, `make reset` can be performed to pull in the latest snapshot. 

Great.  So I think that tells me that if implicit pull (see above) is good enough, we should go with that
 
> > Minor thing: In the documentation for `make reset`, mention that it will not pull in (sync) the configuration committed in `codebase`, and will use the configuration present on the snapshot image only.
> 
> Sure. Note that none of our targets pull in (sync) configuration. So this target behaves as all of our other targets.

Right now, it's "as expected" or "a gotcha", depending on one's intent.  With a little experience under the team's belt, we'll see what works best for us re: documentation or practice.  For now, though, it's a helpful reminder to me.

> > Minor thing: You need to type in a capital Y at the yes/no question, a lower case ends in a make error. Not sure if that's intentional or not
> 
> yes, intentional. That's why it's a capital Y :)

It took me a couple tries to figure that out and confirm that the error I was seeing wasn't something else.  For what it's worth.  Not critical to fix.

**Comment by emetsger** [comment]
I've address the pull feedback.  If there are any other issues, I'd prefer to have a meeting.  Thanks!

**Review by birkland — APPROVED** [review]
This works well with new snapshots now

**Pull request merged**